### PR TITLE
Pipeline to run Exploratory fuzzing tool to catch codegen issues

### PR DESCRIPTION
**Pull request by kunalspathak**
Add an "internal" pipeline to run exploratory tools and return back the repro files.

### CODE REVIEW

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib

File: eng/pipelines/coreclr/exploratory.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# This job definition automates the exploratory tool.
+
+trigger:
+  batch: false
```
Comment (BruceForstall) [inline eng/pipelines/coreclr/exploratory.yml:4]: Do we really want this triggered with every JIT change? That might be useful, but potentially is overkill?

File: eng/pipelines/coreclr/exploratory.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
+# This job definition automates the exploratory tool.
```
Comment (BruceForstall) [inline eng/pipelines/coreclr/exploratory.yml:1]: Should all these be named specifically for the JIT, so it's clear that it's only the JIT being tested? E.g., name this `jit-exploratory.yml`, etc.?

File: eng/pipelines/coreclr/templates/exploratory-job.yml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    # Download coreclr
+    - template: /eng/pipelines/common/download-artifact-step.yml
+      parameters:
+        unpackFolder: $(buildProductRootFolderPath)
+        artifactFileName: '$(buildProductArtifactName)$(archiveExtension)'
+        artifactName: '$(buildProductArtifactName)'
+        displayName: 'Coreclr product build'
+
+    # Download x64 coreclr if running on non-x64 configuration
```
Comment (BruceForstall) [inline eng/pipelines/coreclr/templates/exploratory-job.yml:81]: Why is this necessary? Would it download an x64 product on an ARM32 run, for example?

File: src/coreclr/scripts/exploratory.proj [hunk 4]
```diff
@@ -0,0 +1,87 @@
+<Project Sdk="Microsoft.DotNet.Helix.Sdk" DefaultTargets="Test">
+
+  <!-- <PropertyGroup>
```
Comment (BruceForstall) [inline src/coreclr/scripts/exploratory.proj:3]: You're keeping this, for future ease of debugging? If so, maybe add a comment to that effect. (and how to use it)

File: src/coreclr/scripts/exploratory.proj [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  <PropertyGroup Condition="'$(AGENT_OS)' != 'Windows_NT'">
+    <FileSeparatorChar>/</FileSeparatorChar>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(AGENT_OS)' == 'Windows_NT'">
+    <Python>%HELIX_PYTHONPATH%</Python>
+    <OutputDirectory>%HELIX_WORKITEM_UPLOAD_ROOT%</OutputDirectory>
+    <CoreRoot>%HELIX_CORRELATION_PAYLOAD%\CoreRoot</CoreRoot>
+    <ToolPath>%HELIX_CORRELATION_PAYLOAD%\exploratory</ToolPath>
+    <!-- Workaround until https://github.com/dotnet/arcade/pull/6179 is not available -->
```
Comment (BruceForstall) [inline src/coreclr/scripts/exploratory.proj:31]: "not" available? https://github.com/dotnet/arcade/pull/6179 was merged a year ago, so can this be cleaned up?

File: src/coreclr/scripts/exploratory.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+parser.add_argument("-exploratory_directory", help="Path to exploratory tool")
+parser.add_argument("-output_directory", help="Path to output directory")
+parser.add_argument("-partition", help="Partition name")
+parser.add_argument("-core_root", help="path to CORE_ROOT directory")
+parser.add_argument("-tool_name", help="tool name")
+is_windows = platform.system() == "Windows"
+
+
+def setup_args(args):
+    """ Setup the args for SuperPMI to use.
```
Comment (BruceForstall) [inline src/coreclr/scripts/exploratory.py:39]: There's obviously some copy/paste going on here.

File: src/coreclr/scripts/exploratory.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+################################################################################
+
+import shutil
+import argparse
+from os import linesep, listdir, path, walk
+from os.path import isfile, join, getsize
+import os
+from os import listdir
+from coreclr_arguments import *
+from superpmi_setup import run_command, copy_directory, set_pipeline_variable
```
Comment (BruceForstall) [inline src/coreclr/scripts/exploratory.py:24]: It's worrying that you're importing from superpmi scripts. Maybe there should be some "utility" scripts that both superpmi and exploratory can import from.

**Review by BruceForstall — COMMENTED** [review]
Some comments/questions

There's a lot of code here, and not much in the way of "overall" documentation besides just reading all the code. I worry about that, in terms of maintainability. However, I presume this is just a first step, and you want to get something in and then work on it as we gain more experience with it, and add additional exploratory tools.

File: eng/pipelines/coreclr/exploratory.yml (hunk 1 above)
Comment (kunalspathak) [reply eng/pipelines/coreclr/exploratory.yml:4]: Sounds reasonable. We might not even have to triage the incoming issues. Changed it to trigger nightly and will readjust depending on how useful this proves.

File: eng/pipelines/coreclr/exploratory.yml (hunk 2 above)
Comment (kunalspathak) [reply eng/pipelines/coreclr/exploratory.yml:1]: Sure.

File: eng/pipelines/coreclr/templates/exploratory-job.yml (hunk 3 above)
Comment (kunalspathak) [reply eng/pipelines/coreclr/templates/exploratory-job.yml:81]: Honestly, I don't recall why this was added (this piece is copied from superpmi pipeline). I will try to remove it and see if it breaks anything.

File: src/coreclr/scripts/exploratory.proj (hunk 4 above)
Comment (kunalspathak) [reply src/coreclr/scripts/exploratory.proj]: Deleted it as similar comment is present in other proj files.

File: src/coreclr/scripts/exploratory.proj (hunk 5 above)
Comment (kunalspathak) [reply src/coreclr/scripts/exploratory.proj:15]: This is fixed but I still faced some problems. I will double check.

File: src/coreclr/scripts/exploratory.py (hunk 6 above)
Comment (kunalspathak) [reply src/coreclr/scripts/exploratory.py]: fixed.

File: src/coreclr/scripts/exploratory.py (hunk 7 above)
Comment (kunalspathak) [reply src/coreclr/scripts/exploratory.py]: Yes...eventually as we discussed in https://github.com/dotnet/runtime/pull/53750/files#r646958941. I have opened https://github.com/dotnet/runtime/issues/59559 to track it.

**Comment by kunalspathak** [comment]
> There's a lot of code here, and not much in the way of "overall" documentation 

I have included `Antigen.md` file.

File: src/coreclr/scripts/exploratory.proj (hunk 5 above)
Comment (kunalspathak) [reply src/coreclr/scripts/exploratory.proj:15]: Alright, if i remove that line we see this errors and that's the reason I had it in first place. @safern 

```
D:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\7.0.0-beta.21463.4\tools\download-results\DownloadFromResultsContainer.targets(24,5): error MSB4018: The "DownloadFromResultsContainer" task failed unexpectedly. [D:\workspace\_work\1\s\src\coreclr\scripts\exploratory.proj]
D:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\7.0.0-beta.21463.4\tools\download-results\DownloadFromResultsContainer.targets(24,5): error MSB4018: System.IO.IOException: The filename, directory name, or volume label syntax is incorrect. : 'D:\workspace\_work\1\s\src\coreclr\scripts\$([MSBuild]::NormalizeDirectory('$(BUILD_SOURCESDIRECTORY)', 'artifacts', helixresults'))\9e889cb3-8222-4f97-b8c4-c95ff9d8fa88' [D:\workspace\_work\1\s\src\coreclr\scripts\exploratory.proj]
D:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\7.0.0-beta.21463.4\tools\download-results\DownloadFromResultsContainer.targets(24,5): error MSB4018:    at System.IO.FileSystem.CreateDirectory(String fullPath, Byte[] securityDescriptor) [D:\workspace\_work\1\s\src\coreclr\scripts\exploratory.proj]
D:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\7.0.0-beta.21463.4\tools\download-results\DownloadFromResultsContainer.targets(24,5): error MSB4018:    at System.IO.Directory.CreateDirectory(String path) [D:\workspace\_work\1\s\src\coreclr\scripts\exploratory.proj]
D:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\7.0.0-beta.21463.4\tools\download-results\DownloadFromResultsContainer.targets(24,5): error MSB4018:    at Microsoft.DotNet.Helix.Sdk.DownloadFromResultsContainer.ExecuteCore() in /_/src/Microsoft.DotNet.Helix/Sdk/DownloadFromResultsContainer.cs:line 66 [D:\workspace\_work\1\s\src\coreclr\scripts\exploratory.proj]
D:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\7.0.0-beta.21463.4\tools\download-results\DownloadFromResultsContainer.targets(24,5): error MSB4018:    at Microsoft.DotNet.Helix.Sdk.DownloadFromResultsContainer.Execute() in /_/src/Microsoft.DotNet.Helix/Sdk/DownloadFromResultsContainer.cs:line 58 [D:\workspace\_work\1\s\src\coreclr\scripts\exploratory.proj]
D:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\7.0.0-beta.21463.4\tools\download-results\DownloadFromResultsContainer.targets(24,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [D:\workspace\_work\1\s\src\coreclr\scripts\exploratory.proj]
D:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\7.0.0-beta.21463.4\tools\download-results\DownloadFromResultsContainer.targets(24,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [D:\workspace\_work\1\s\src\coreclr\scripts\exploratory.proj]
##[error].packages\microsoft.dotnet.helix.sdk\7.0.0-beta.21463.4\tools\download-results\DownloadFromResultsContainer.targets(24,5): error MSB4018: (NETCORE_ENGINEERING_TELEMETRY=Build) The "DownloadFromResultsContainer" task failed unexpectedly.
System.IO.IOException: The filename, directory name, or volume label syntax is incorrect. : 'D:\workspace\_work\1\s\src\coreclr\scripts\$([MSBuild]::NormalizeDirectory('$(BUILD_SOURCESDIRECTORY)', 'artifacts', helixresults'))\9e889cb3-8222-4f97-b8c4-c95ff9d8fa88'
   at System.IO.FileSystem.CreateDirectory(String fullPath, Byte[] securityDescriptor)
   at System.IO.Directory.CreateDirectory(String path)
   at Microsoft.DotNet.Helix.Sdk.DownloadFromResultsContainer.ExecuteCore() in /_/src/Microsoft.DotNet.Helix/Sdk/DownloadFromResultsContainer.cs:line 66
   at Microsoft.DotNet.Helix.Sdk.DownloadFromResultsContainer.Execute() in /_/src/Microsoft.DotNet.Helix/Sdk/DownloadFromResultsContainer.cs:line 58
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)

```

File: eng/pipelines/coreclr/exploratory.yml (hunk 1 above)
Comment (safern) [inline eng/pipelines/coreclr/exploratory.yml:4]: should we start with nightly runs only?

File: eng/pipelines/coreclr/exploratory.yml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+# This pipeline is supposed to be run only on merged changes
+# and should not be triggerable from a PR. 
+pr: none
+
+schedules:
+- cron: "0 17 * * 0"
+  displayName: Sun at 9:00 AM (UTC-8:00)
+  branches:
+    include:
+    - main
```
Comment (safern) [inline eng/pipelines/coreclr/exploratory.yml:22]: can we filter by paths on schedules? (I don't know if it is possible).

File: eng/pipelines/coreclr/templates/exploratory-job.yml [hunk 9]
```diff
@@ -0,0 +1,93 @@
+parameters:
+  buildConfig: ''
+  archType: ''
+  osGroup: ''
+  osSubgroup: ''
+  container: ''
+  runtimeVariant: ''
+  testGroup: ''
+  framework: net5.0 # Specify the appropriate framework when running release branches (ie netcoreapp3.0 for release/3.0)
```
Comment (safern) [inline eng/pipelines/coreclr/templates/exploratory-job.yml:9]: why is  the framework defaulting to net5.0?

File: eng/pipelines/coreclr/exploratory.yml (hunk 1 above)
Comment (kunalspathak) [reply eng/pipelines/coreclr/exploratory.yml:4]: yes, I changed it to nightly.

File: eng/pipelines/coreclr/templates/exploratory-job.yml (hunk 9 above)
Comment (kunalspathak) [reply eng/pipelines/coreclr/templates/exploratory-job.yml:9]: let me change to net6.0. Thanks for pointing.

File: eng/pipelines/coreclr/templates/run-exploratory-job.yml [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+  jobName: ''                     # required -- job name
+  displayName: ''                 # optional -- display name for the job. Will use jobName if not passed
+  pool: ''                        # required -- name of the Build pool
+  container: ''                   # required -- name of the container
+  buildConfig: ''                 # required -- build configuration
+  archType: ''                    # required -- targeting CPU architecture
+  osGroup: ''                     # required -- operating system for the job
+  osSubgroup: ''                  # optional -- operating system subgroup
+  extraSetupParameters: ''        # optional -- extra arguments to pass to the setup script
+  frameworks: ['netcoreapp3.0']   # optional -- list of frameworks to run against
```
Comment (safern) [inline eng/pipelines/coreclr/templates/run-exploratory-job.yml:13]: this is defaulting to a framework that is EOL already, can we update this?

File: src/coreclr/scripts/exploratory.proj (hunk 5 above)
Comment (safern) [reply src/coreclr/scripts/exploratory.proj:15]: It seems like helixresults in that call to `NormalizeDirectory` is missing an opening `'`?

Comment (kunalspathak) [reply src/coreclr/scripts/exploratory.proj:15]: will you be able to send a fix? I won't block this PR, but will send follow-up PR to validate and remove this.

File: eng/pipelines/coreclr/exploratory.yml (hunk 8 above)
Comment (kunalspathak) [reply eng/pipelines/coreclr/exploratory.yml:22]: Just changed it to nightly so don't need to filter by paths I suppose?

File: src/coreclr/scripts/exploratory.proj (hunk 5 above)
Comment (safern) [reply src/coreclr/scripts/exploratory.proj:15]: Would you be able to validate that fixes the issue locally? You can update the targets locally from the nuget cache and then send the job from your local box and see if it works?

File: eng/pipelines/coreclr/templates/run-exploratory-job.yml (hunk 10 above)
Comment (kunalspathak) [reply eng/pipelines/coreclr/templates/run-exploratory-job.yml:13]: sure.

File: src/coreclr/scripts/exploratory.proj (hunk 5 above)
Comment (kunalspathak) [reply src/coreclr/scripts/exploratory.proj:15]: I was not able to verify this on local machine and probably take me time, but looking at other references (e.g. https://github.com/dotnet/runtime/blob/2e08554621bcd8415bccd581e7be06b48c7813d8/src/mono/wasm/build/WasmApp.targets#L167), I think `$(BUILD_SOURCESDIRECTORY)` doesn't need the surrounding `'`'. I will send a PR to arcade.

Comment (kunalspathak) [reply src/coreclr/scripts/exploratory.proj:15]: Sent out https://github.com/dotnet/arcade/pull/7961

File: eng/pipelines/coreclr/templates/exploratory-job.yml (hunk 3 above)
Comment (kunalspathak) [reply eng/pipelines/coreclr/templates/exploratory-job.yml]: Removed it.

File: eng/pipelines/coreclr/exploratory.yml [hunk 11]
```diff
@@ -0,0 +1,51 @@
+# This job definition automates the exploratory tool.
+
+trigger: none
+
+schedules:
+- cron: "0 6 * * *"
+  displayName: Daily at 11:30 PM (UTC-8:00)
```
Comment (BruceForstall) [inline eng/pipelines/coreclr/exploratory.yml:7]: It seems like daily is too often, and weekly might make more sense. But I'm ok with this; we can change our mind about scheduling later.

File: eng/pipelines/common/download-specific-artifact-step.yml [hunk 12]
```diff
@@ -12,10 +12,10 @@ steps:
     inputs:
       buildType: specific
       project:  'internal' # 'internal' or 'public'
-      pipeline: 'superpmi-replay'
```
Comment (BruceForstall) [inline eng/pipelines/common/download-specific-artifact-step.yml:15]: Is this file only used for testing?

File: src/coreclr/scripts/Antigen.md [hunk 13]
```diff
@@ -0,0 +1,23 @@
+# Documentation of Exploratory tool Antigen
```
Comment (BruceForstall) [inline src/coreclr/scripts/Antigen.md:1]: nit: filenames should be all lower case to make working on Linux easier.

File: src/coreclr/scripts/antigen_setup.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+    antigen_dst_directory = path.join(correlation_payload_directory, "exploratory")
+
+    helix_source_prefix = "official"
+    creator = ""
+    ci = True
+    if is_windows:
+        helix_queue = "Windows.10.Arm64" if arch_name == "arm64" else "Windows.10.Amd64.X86"
+    else:
+        if arch_name == "arm":
+            helix_queue = "(Ubuntu.1804.Arm32)Ubuntu.1804.Armarch@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm32v7-bfcd90a-20200121150440"
```
Comment (BruceForstall) [inline src/coreclr/scripts/antigen_setup.py:96]: It seems problematic that we have Helix queue names hard-coded here, as well as in superpmi_setup.py, instead of using the centralized helix-queues-setup.yml file. If nothing else, the python scripts should share a small, simple, helix-queues-setup.py helper script if we can't/don't use the yml file. This change could be done as a follow-up.

File: src/coreclr/scripts/antigen_setup.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+                if not is_windows:
+                    dotnet_cmd = os.path.join(source_directory, "dotnet.sh")
+                run_command([dotnet_cmd, "publish", "-c", "Release", "--self-contained", "-r", run_configuration, "-o", antigen_bin_directory], _exit_on_fail=True)
+
+            if not os.path.exists(path.join(antigen_bin_directory, "Antigen.dll")):
+                raise FileNotFoundError("Antigen.dll not present at {}".format(antigen_bin_directory))
+
+            # copy antigen tool
+            print('Copying {} -> {}'.format(antigen_bin_directory, antigen_dst_directory))
+            copy_directory(antigen_bin_directory, antigen_dst_directory, match_func=acceptable_copy)
```
Comment (BruceForstall) [inline src/coreclr/scripts/antigen_setup.py:136]: I wonder if this should all be encapsulated in a script in the antigen source directory, so this script doesn't need to know how to build the tool. Something like `<tool_code_directory>build.cmd <dotnet_tool/script_to_use> <antigen_dst_directory>`.

**Review by BruceForstall — APPROVED** [review]
LGTM. A couple comments for future considertion.

File: eng/pipelines/coreclr/exploratory.yml (hunk 11 above)
Comment (kunalspathak) [reply eng/pipelines/coreclr/exploratory.yml:7]: Ok, I will schedule it for weekly for now.

File: eng/pipelines/common/download-specific-artifact-step.yml (hunk 12 above)
Comment (kunalspathak) [reply eng/pipelines/common/download-specific-artifact-step.yml:15]: Yes, this is used to skip building coreclr/libraries and download these from previous runs. This helps in fast testing during pipeline development.

File: src/coreclr/scripts/antigen_setup.py (hunk 14 above)
Comment (kunalspathak) [reply src/coreclr/scripts/antigen_setup.py:96]: Sounds reasonable. Added in the list of https://github.com/dotnet/runtime/issues/59559

File: src/coreclr/scripts/antigen_setup.py (hunk 15 above)
Comment (kunalspathak) [reply src/coreclr/scripts/antigen_setup.py:136]: Sure, I will fix it in antigen and make related change here as a follow-up PR.
